tests/utils/test_schemapi.py

```diff
@@ -1024,3 +1025,139 @@ def test_to_dict_range(tp) -> None:
+        ),
+        (
+            (
+                dt.datetime(2003, 5, 1, 1, 30, tzinfo=dt.timezone.utc),
+                dt.datetime(2003, 6, 3, 4, 3, tzinfo=dt.timezone.utc),
+            ),
+            (
+                DateTime(2003, 5, 1, 1, 30, 0, 0, utc=True),
+                DateTime(2003, 6, 3, 4, 3, 0, 0, utc=True),
+            ),
+        ),
+    ],
+    ids=["date", "datetime (no time)", "datetime (microseconds)", "datetime (UTC)"],
+)
+def test_to_dict_datetime(
+    stocks, window: tuple[Any, Any], expected: tuple[alt.DateTime, alt.DateTime]
+) -> None:
+    """
+    Includes `datetime.datetime` with an empty time component.
+
+    This confirms that conversion matches how `alt.DateTime` omits `Undefined`.
+    """
+    expected_dicts = [e.to_dict() for e in expected]
+    brush = alt.selection_interval(encodings=["x"], value={"x": window})
+    base = stocks
+
+    upper = base.encode(alt.X("date:T").scale(domain=brush))
+    lower = base.add_params(brush)
+    chart = upper & lower
+    mapping = chart.to_dict()
+    params_value = mapping["params"][0]["value"]["x"]
+
+    assert isinstance(params_value, list)
+    assert params_value == expected_dicts
+
+
+@pytest.mark.parametrize(
+    "tzinfo",
+    [
+        dt.timezone(dt.timedelta(hours=2), "UTC+2"),
+        dt.timezone(dt.timedelta(hours=1), "BST"),
+        dt.timezone(dt.timedelta(hours=-7), "pdt"),
+        dt.timezone(dt.timedelta(hours=-3), "BRT"),
+    ],
+)
+def test_to_dict_datetime_unsupported_timezone(tzinfo: dt.timezone) -> None:
+    datetime = dt.datetime(2003, 5, 1, 1, 30)
+
+    result = alt.FieldEqualPredicate(datetime, "column 1")  # type: ignore[arg-type]
+    assert result.to_dict()
+
+    with pytest.raises(TypeError, match=r"Unsupported timezone.+\n.+UTC.+local"):
+        alt.FieldEqualPredicate(datetime.replace(tzinfo=tzinfo), "column 1")  # type: ignore[arg-type]
+
+
+def test_to_dict_datetime_typing() -> None:
+    """
+    Enumerating various places that need updated annotations.
+
+    All work at runtime, just need to give the type checkers the new info.
+
+    Sub-issue of https://github.com/vega/altair/issues/3650
+    """
+    datetime = dt.datetime(2003, 5, 1, 1, 30)
+    datetime_seq = [datetime, datetime.replace(2005), datetime.replace(2008)]
+    assert alt.FieldEqualPredicate(datetime, field="column 1")  # type: ignore[arg-type]
+    assert alt.FieldOneOfPredicate(
+        oneOf=datetime_seq,  # type: ignore[arg-type]
+        field="column 1",
+    )
+
+    assert alt.Legend(values=datetime_seq)  # type: ignore[arg-type]
+
+    assert alt.Scale(domain=datetime_seq)  # type: ignore[arg-type]
+    assert alt.Scale(domainMin=datetime_seq[0], domainMax=datetime_seq[2])  # type: ignore[arg-type]
+
+    # NOTE: `datum` is not annotated?
+    assert alt.XDatum(datum=datetime).to_dict()
+
+    # NOTE: `*args` is not annotated?
+    # - All of these uses *args incorrectly
+    assert alt.Vector2DateTime(datetime_seq[:2])
```

Comment:
# TODO
Ensure all of these ignore comments are marked as unused

Reply:
Resolved in: [fix(typing): Remove unused ignores](https://github.com/vega/altair/pull/3653/commits/29c500c925974b2ed672554e644a10af212a9d42)